### PR TITLE
Add WebCodecs article.

### PR DESCRIPTION
**Pull request by jpmedley**
Address #3695

### CODE REVIEW

File: src/site/content/en/blog/web-codecs/index.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+  {
+    method: 'POST',
+    headers: {   'Content-Type': 'application/octet-stream'   },
+    body: data
+  });
+}
+```
+
+If at some point you'd need to make sure that all pending encoding requests have
+been completed, you can call `flush()` and wait for its promise. For example,
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:203]:
```suggestion
been completed, you can call `flush()` and wait for its promise.
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+it's required to call `flush()` before the encoder can be reconfigured.
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:204]:
```suggestion
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+This is an easy way to check for WebCodecs support, to make  sure that the
+origin trial or the flag took effect.
+
+```js
+if ("VideoEncoder" in window) {
+  // WebCodecs API is supported.
+}
+```
+
+## Using the WebHID API {: #use }
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:339]:
```suggestion
## Using the WebCodecs API {: #use }
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+### Enabling via chrome://flags
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:341]: There is no way to enable WebCodecs via chrome://flags. 
Please preserve original wording:

```
To enable WebCodecs locally, start Chrome with a command line flag: --enable-blink-features=WebCodecs.
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+To experiment with the WebHID API locally on all desktop platforms, without an
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:343]:
```suggestion
To experiment with the WebCodecs API locally on all desktop platforms, without an
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+  bitrate: 8_000_000,     // 8 Mbps
+  framerate: 30,
+};
+
+let encoder = new VideoEncoder(init);
+encoder.configure(config);
+```
+
+After the encoder has been set up,  it's ready to start accepting frames. When
+frames are coming from a media stream, `VideoTrackReader` will pump frames into
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:148]:
```suggestion
frames are coming from a media stream, the `VideoTrackReader` callback will pump frames into
```

File: src/site/content/en/blog/web-codecs/index.md [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+```
+
+### Enabling support during the origin trial phase
+
+The WebCodecs API is available on all desktop platforms (Chrome OS, Linux, macOS,
+and Windows) as an origin trial in Chrome&nbsp;86. The origin trial is expected
+to end just before Chrome&nbsp;88 moves to stable in February 2021. The API can
+also be enabled using a flag.
+
+{% include 'content/origin-trials.njk' %}
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:357]: Does this template provide a link to the right origin trial?

File: src/site/content/en/blog/web-codecs/index.md [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+feedback:
+  - api
+---
+
+Modern web technologies provide ample ways to work with video.
+[Media Stream API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),
+[Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API)
+,
+[Media Source API](https://developer.mozilla.org/en-US/docs/Web/API/Media_Source_Extensions_API),
+[WebRTC API](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API) add up
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:25]:
```suggestion
 and [WebRTC API](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API) add up
```

**Review by Djuffin — COMMENTED** [review]
It looks like the feedback section got lost.

File: src/site/content/en/blog/web-codecs/index.md (hunk 6 above)
Comment (jpmedley) [reply src/site/content/en/blog/web-codecs/index.md:148]: What is he trying to communicate? `VideoTrackReader` is an interface, not a callback.

Comment (jpmedley) [reply src/site/content/en/blog/web-codecs/index.md:148]: Is he trying to refer to the callback passed to `VideoTrackReader.start()`? If so, it needs to be that explicit. What he suggested is too big of a leap for people don't know anything about WebCodecs (which is anyone reading this article).

File: src/site/content/en/blog/web-codecs/index.md (hunk 7 above)
Comment (jpmedley) [reply src/site/content/en/blog/web-codecs/index.md:356]: Yes. It reads the URL from the metadata at the top.

File: src/site/content/en/blog/web-codecs/index.md (hunk 6 above)
Comment (Djuffin) [reply src/site/content/en/blog/web-codecs/index.md:148]: Yeah, that's exactly what he means here.  How about this  
```suggestion
frames are coming from a media stream, the callback given to `VideoTrackReader.start()` will pump frames into
```

Feel free to reword it in whatever way you see fit.

File: src/site/content/en/blog/web-codecs/index.md (hunk 4 above)
Comment (jpmedley) [reply src/site/content/en/blog/web-codecs/index.md]: Fixed.

File: src/site/content/en/blog/web-codecs/index.md [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+
+## Using the WebCodecs API {: #use }
+
+### Enabling via a command line flag
+
+To experiment with the WebCodecs API locally on all desktop platforms, without an
+origin trial token, start Chrome with a command line flag:
+
+```bash
+--enable-blink-features=WebCodecs.
```
Comment (beaufortfrancois) [inline src/site/content/en/blog/web-codecs/index.md:360]:
```suggestion
--enable-blink-features=WebCodecs

**Review by Djuffin — COMMENTED** [review]
I like the complete drawing diagram. Are you planning on inserting it anywhere in the post?

**Comment by Djuffin** [comment]
This pull request has Chrome 87 label, but it's actually about Chrome 86

**Comment by Djuffin** [comment]
Fair enough.
> The job of an article like this is to present untangled. Readers shouldn't be asked to untangle and understand at the same time.

**Comment by Djuffin** [comment]
Never mind.
I was referring to the red label (Chrome 87) on the pull request itself, not in the text. But I think it just marks the due date for the blog post, not the an Chrome release.  
> It [does not appear so now](https://github.com/GoogleChrome/web.dev/pull/4022/files#diff-23546551ebbf228a95b62f2c3444ed9a521a3fadde452546c167274ea3f5ca6dR366).

**Review by kaycebasques — CHANGES_REQUESTED** [review]
- [ ] Currently the URL is `web.dev/web-codecs` yet the API name is WebCodecs. It seems like we should therefore change the URL to `web.dev/webcodecs`?
- [ ] The `alt` descriptions on the images are all the same. Please update them to properly describe each diagram.
- [ ] The [Current status](https://deploy-preview-4022--web-dev-staging.netlify.app/web-codecs/#status) section links to a Feedback section yet no Feedback section exists.
- [ ] Is there a reason we weren't embedding the Glitch into the page? I changed it to an embed.
- [ ] Can/should we use [code highlighting](https://web.dev/handbook/markup-code/#code-highlighting) to call out any specific sections of the code samples?

**Comment by jpmedley** [comment]
> It seems like we should therefore change the URL to web.dev/webcodecs?

Aren't we required to have a dash in the URL?

>The alt descriptions on the images are all the same. Please update them to properly describe each diagram.

Please look more carefully.

>The Current status section links to a Feedback section yet no Feedback section exists.

Feedback added.

>Is there a reason we weren't embedding the Glitch into the page? I changed it to an embed.

No reason. I just don't personally 

>Can/should we use code highlighting to call out any specific sections of the code samples?

I've only found one place where that's needed, but it's done.

File: src/site/content/en/blog/web-codecs/index.md [hunk 10]
```diff
@@ -373,6 +373,36 @@ also be enabled using a flag.
 
 {% include 'content/origin-trial-register.njk' %}
 
+## Feedback {: #feedback }
+
+The Chrome team wants to hear about your experiences with the Idle Detection API.
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:378]: WebCodecs

File: src/site/content/en/blog/web-codecs/index.md [hunk 11]
```diff
@@ -373,6 +373,36 @@ also be enabled using a flag.
...
+
+Did you find a bug with Chrome's implementation? Or is the implementation
+different from the spec? File a bug at [new.crbug.com](https://new.crbug.com).
+Be sure to include as much detail as you can, simple instructions for
+reproducing, and enter `Blink>Media>WebCodecs` in the **Components** box.
+[Glitch](https://glitch.com/) works great for sharing quick and easy repros.
+
+### Show support for the API
+
+Are you planning to use the Idle Detection API? Your public support helps the
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:398]: WebCodecs

File: src/site/content/en/blog/web-codecs/index.md (hunk 10 above)
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:378]:
```suggestion
The Chrome team wants to hear about your experiences with the WebCodecs API.
```

File: src/site/content/en/blog/web-codecs/index.md (hunk 11 above)
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:398]: WebCodecs

File: src/site/content/en/blog/web-codecs/index.md [hunk 12]
```diff
@@ -373,6 +373,36 @@ also be enabled using a flag.
...
+Chrome team to prioritize features and shows other browser vendors how critical
+it is to support them.
+
+Share how you plan to use it on the [WICG Discourse thread][wicg-discourse] Send
+a Tweet to [@ChromiumDev][cr-dev-twitter] with the `#idledetection` hashtag and
```
Comment (Djuffin) [inline src/site/content/en/blog/web-codecs/index.md:403]:
```suggestion
a Tweet to [@ChromiumDev][cr-dev-twitter] with the `#webcodecs` hashtag and
```

**Comment by kaycebasques** [comment]
@jpmedley FYI

We are not required to have a dash in our pathnames. LMK where you got that from and I will correct it.

Here are the images that still have the same `alt` descriptions: 

image

image

I also thought this one had the same `alt` but I see now I was mistaken

image

File: src/site/content/en/blog/web-codecs/index.md [hunk 13]
```diff
@@ -0,0 +1,408 @@
+---
+title: Video processing with WebCodecs
+subhead: Manipulating video stream components.
+description: |
+  Work with components of a video stream, such as frames and unmuxed chunks of encoded video or audio.
+date: 2020-10-13
+updated: 2020-10-13
```
Comment (kaycebasques) [inline src/site/content/en/blog/web-codecs/index.md:7]:
```suggestion
# updated: 2020-10-13
```
I usually omit this until we actually update the page

File: src/site/content/en/blog/web-codecs/index.md [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+  - media
+  - video
+feedback:
+  - api
+---
+
+Modern web technologies provide ample ways to work with video.
+[Media Stream API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),
+[Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API)
+,
```
Comment (kaycebasques) [inline src/site/content/en/blog/web-codecs/index.md:25]:
```suggestion
[Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),
```
Nitpick

File: src/site/content/en/blog/web-codecs/index.md [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+let frames_from_stream = [];
+let stream = await navigator.mediaDevices.getUserMedia({ … });
+let vtr = new VideoTrackReader(stream.getVideoTracks()[0]);
+vtr.start((frame) => {
+  frames_from_stream.push(frame);
+});
+```
+
+<figure class="w-figure">
+  <img src="./consuming-frames-from-tracks.png" alt="The path from ImageBitmap to the network or to storage.">
```
Comment (kaycebasques) [inline src/site/content/en/blog/web-codecs/index.md:122]:
```suggestion
  <img src="./consuming-frames-from-tracks.png" alt="The path from MediaStreamTrack to the network or to storage.">
```

File: src/site/content/en/blog/web-codecs/index.md (hunk 11 above)
Comment (kaycebasques) [reply src/site/content/en/blog/web-codecs/index.md:398]:
```suggestion
Are you planning to use the WebCodecs API? Your public support helps the
```

**Review by kaycebasques — CHANGES_REQUESTED** [review]

**Review by kaycebasques — APPROVED** [review]

**Pull request merged**